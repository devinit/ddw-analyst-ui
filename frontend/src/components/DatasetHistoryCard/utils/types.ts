export interface SavedQueryData {
  id: number;
  saved_query_db_table: string;
  operation: number;
  full_query: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
}
