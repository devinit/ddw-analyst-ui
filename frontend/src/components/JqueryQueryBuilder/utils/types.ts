export type JqueryQueryBuilderFilter = {
  condition?: string;
  rules?: (JqueryQueryBuilderFilterComparator | JqueryQueryBuilderFilter)[];
};

export type JqueryQueryBuilderHaving = {
  condition?: string;
  rules?: (JqueryQueryBuilderHavingComparator | JqueryQueryBuilderHaving)[];
};

export type JqueryQueryBuilderHavingComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value?: string | number | { plain: string | number } | { column: string; aggregate: string };
  values?: { value: string; label: string }[];
};

export type JqueryQueryBuilderFilterComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value: string | number;
};
export type JqueryQueryBuilderComps =
  | 'equal'
  | 'not_equal'
  | 'less'
  | 'less_or_equal'
  | 'greater'
  | 'greater_or_equal'
  | 'between'
  | 'in';

export type JqueryQueryBuilderFieldData = {
  id: string | number;
  label: string;
  type: string;
};

export type JqueryQueryBuilderIcons = {
  add_group: string;
  add_rule: string;
  remove_group: string;
  remove_rule: string;
  error: string;
};
