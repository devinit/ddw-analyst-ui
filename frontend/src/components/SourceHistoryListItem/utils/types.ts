export interface FrozenData {
  id: number;
  parent_db_table: string;
  frozen_db_table: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
  logs: string | null;
}
