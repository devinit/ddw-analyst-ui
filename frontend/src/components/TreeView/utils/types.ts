import { ReactElement } from 'react';

export interface TreeViewProps {
  data: Data[];
  depth?: number;

  deleteElement?: ReactElement;

  getStyleClassCb?: (node: Data) => void;

  isCheckable?: (node: Data, depth?: number) => void;
  isDeletable?: (node: Data, depth?: number) => void;
  isExpandable?: (node: Data, depth?: number) => void;

  keywordChildren?: string;
  keywordChildrenLoading?: string;
  keywordKey?: string;
  keywordLabel?: string;

  loadingElement?: ReactElement;
  noChildrenAvailableMessage?: string;

  onCheckToggleCb?: (data: Data[], depth?: number) => void;
  onDeleteCb?: (node: Data, data: Data[], depth?: number) => void;
  onExpandToggleCb?: (node: Data, depth?: number) => void;
  onUpdateCb?: (data: unknown[], depth?: number) => void;

  transitionEnterTimeout?: number;
  transitionExitTimeout?: number;
}

export interface Data {
  id: number;
  name: string;
  isExpanded?: boolean;
  isChecked?: boolean;
  isChildrenLoading?: boolean;
  children?: Data[];
}
