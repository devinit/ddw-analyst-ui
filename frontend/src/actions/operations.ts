import {
  FETCH_OPERATION,
  FETCH_OPERATIONS,
  FETCH_OPERATION_FAILED,
  OperationsAction,
  SET_ACTIVE_OPERATION
} from '../reducers/operations';
import { OperationMap } from '../types/operations';

interface FetchOptions {
  limit?: number;
  offset?: number;
  link?: string;
}
export const fetchOperations = ({ limit = 10, offset = 0, link }: FetchOptions): Partial<OperationsAction> =>
  ({ type: FETCH_OPERATIONS, payload: { limit, offset, link } });
export const fetchOperation = (id: string): Partial<OperationsAction> => ({ type: FETCH_OPERATION, payload: { id } });
export const setOperation = (activeOperation?: OperationMap, loading = false): Partial<OperationsAction> =>
  ({ type: SET_ACTIVE_OPERATION, activeOperation, loading });
export const fetchOperationFailed = (): Partial<OperationsAction> => ({ type: FETCH_OPERATION_FAILED });
