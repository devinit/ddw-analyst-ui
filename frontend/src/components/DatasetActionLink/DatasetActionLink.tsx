import React, { FunctionComponent, MouseEvent } from 'react';
import { OperationMap } from '../../types/operations';

interface ComponentProps {
  operation: OperationMap;
  show?: boolean;
  path?: string;
  action?: 'edit' | 'data';
  onClick?: (operation: OperationMap) => void;
}

const DatasetActionLink: FunctionComponent<ComponentProps> = (props) => {
  if (props.show) {
    const onClick = (event: MouseEvent<HTMLAnchorElement>) => {
      event.preventDefault();
      if (props.onClick) {
        props.onClick(props.operation);
      }
    };
    const action = props.action === 'edit' ? 'build' : 'data';
    const href = props.path || `/queries/${action}/${props.operation.get('id') as number}/`;

    return (
      <a className="btn btn-sm btn-dark" href={href} onClick={onClick} data-testid="dataset-action">
        {props.children}
      </a>
    );
  }

  return null;
};

DatasetActionLink.defaultProps = {
  show: true,
  action: 'edit',
};

export { DatasetActionLink };
