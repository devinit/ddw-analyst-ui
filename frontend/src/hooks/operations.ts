import axios, { AxiosResponse } from 'axios';
import { fromJS } from 'immutable';
import * as localForage from 'localforage';
import { useEffect, useState } from 'react';
import { Operation, OperationMap } from '../types/operations';
import { api, localForageKeys } from '../utils';

interface UseOperationResult {
  operation?: OperationMap;
  loading: boolean;
}

export const useOperation = (id?: number, fetch = false): UseOperationResult => {
  if (!id) {
    return { loading: false, operation: undefined };
  }
  const [operation, setOperation] = useState<OperationMap | undefined>(fromJS({}));
  const [loading, setLoading] = useState(false);

  const fetchOperation = () => {
    setLoading(true);
    axios
      .request({
        url: `${api.routes.SINGLE_DATASET}${id}`,
        method: 'get',
        headers: {
          'Content-Type': 'application/json',
        },
      })
      .then(({ status, data, statusText }: AxiosResponse<Operation>) => {
        if (status === 200 && data) {
          const activeOperation = fromJS(data);
          setOperation(activeOperation);
          setLoading(false);
        } else if (status === 401) {
          console.log('Failed to fetch operation: ', statusText);
          setOperation(undefined);
          setLoading(false);
        }
      })
      .catch((error) => {
        console.log(
          `Failed to fetch operation: ${error.response.status} ${error.response.statusText}`,
        );
        setOperation(undefined);
        setLoading(false);
      });
  };

  useEffect(() => {
    if (fetch) {
      fetchOperation();
    } else {
      localForage
        .getItem<Operation | undefined>(localForageKeys.ACTIVE_OPERATION)
        .then((activeOperation) => {
          if (activeOperation && activeOperation.id === id) {
            setOperation(fromJS(activeOperation));
            setLoading(false);
          } else {
            fetchOperation();
          }
        });
    }
  }, [id]);

  return { loading, operation };
};
