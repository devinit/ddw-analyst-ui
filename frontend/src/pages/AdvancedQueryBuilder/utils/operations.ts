import axios from 'axios';
import { Operation, OperationMap } from '../../../types/operations';
import { api } from '../../../utils';
import { clearOperationsCache } from '../../../utils/cache';

export const saveOperation = async (operation: OperationMap, token: string): Promise<void> => {
  if (!operation) {
    return;
  }
  const id = operation.get('id');
  const url = id ? `${api.routes.SINGLE_DATASET}${id}/` : api.routes.DATASETS;
  const data: Operation = operation.toJS() as Operation;
  const response = await axios.request<Operation>({
    url,
    method: id ? 'put' : 'post',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
    data,
  });
  if (response.status === 200 || response.status === 201) {
    clearOperationsCache();
  }

  return;
};

export const deleteOperation = async (operationID: string, token: string): Promise<void> => {
  const response = await axios.request({
    url: `${api.routes.SINGLE_DATASET}${operationID}/`,
    method: 'delete',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
  });

  if (response.status === 200 || response.status === 204 || response.status === 201) {
    clearOperationsCache();

    return;
  }

  throw new Error(
    `Operation delete failed: Code: ${response.status} | Message: ${response.statusText}`,
  );
};
