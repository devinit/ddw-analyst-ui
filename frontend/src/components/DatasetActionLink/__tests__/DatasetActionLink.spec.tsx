/**
 * @jest-environment jsdom
 */
import { fireEvent, render, screen } from '@testing-library/react';
import { List, Map } from 'immutable';
import 'jest-styled-components';
import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { Operation, OperationMap } from '../../../types/operations';
import { DatasetActionLink } from '../DatasetActionLink';

const operation: Partial<Operation> = {
  id: 23,
  name: 'My Test Dataset',
  description: 'My Dataset Description',
  updated_on: new Date('2020-01-02').toISOString(),
  is_draft: false,
  operation_steps: List(),
};
const props = {
  operation: Map(operation) as OperationMap,
};

describe('DatasetActionLink', () => {
  test('renders correctly with the default props', () => {
    const renderer = TestRenderer.create(<DatasetActionLink {...props} />).toJSON();

    expect(renderer).toMatchSnapshot();
  });

  test('does not render a link when show is false', () => {
    const renderer = TestRenderer.create(<DatasetActionLink {...props} show={false} />).toJSON();

    expect(renderer).toBeNull();
  });

  test('renders with the provided path', () => {
    const path = '/dataset/action/link';
    render(<DatasetActionLink {...props} path={path} />);
    const link = screen.getByTestId('dataset-action');

    expect(link.getAttribute('href')).toBe(path);
  });

  test('with no path and action edit renders the appropriate link', () => {
    render(<DatasetActionLink {...props} action="edit" />);
    const link = screen.getByTestId('dataset-action');

    expect(link.getAttribute('href')).toBe('/queries/build/23/');
  });

  test('responds to clicks', () => {
    const onClick = jest.fn();
    render(<DatasetActionLink {...props} onClick={onClick} />);
    fireEvent.click(screen.getByTestId('dataset-action'));

    expect(onClick).toHaveBeenCalled();
  });
});
