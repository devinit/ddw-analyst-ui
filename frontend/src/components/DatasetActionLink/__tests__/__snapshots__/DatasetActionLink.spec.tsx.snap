// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetActionLink renders correctly with the default props 1`] = `
<a
  className="btn btn-sm btn-dark"
  data-testid="dataset-action"
  href="/queries/build/23/"
  onClick={[Function]}
/>
`;
